### PR TITLE
Move "cirq_type" responsibility to protocols

### LINKED ISSUE #4698: Enforce [namespace.]classname format for cirq_type
The type-serialization behavior added in #4693 made an implicit assumption that Cirq objects use `object.__class__.__name__` for the "cirq_type" field of their JSON representation, optionally prepended with a namespace. This is consistent with existing Cirq types, but until that PR it was merely a convention; now that code depends on it, we ought to enforce it.

More specifically, this issue recommends moving assignment of the `cirq_type` JSON field from the `_json_dict_` magic method of serializable classes into the `cirq.to_json` protocol. As discussed in the linked issue, this will provide consistent behavior and reduce code redundancy.

### PR DESCRIPTION
**Pull request by 95-martin-orion**
Fixes #4698.

This enforces the `'cirq_type': [<namespace>.]<classname>` convention for JSON serialization by assigning the "cirq_type" field in our protocols instead of in each class. Any external users who define a "cirq_type" will see a warning about this; after a deprecation cycle, we will begin raising errors to prevent explicit definition of "cirq_type" in classes.

### CODE REVIEW

**Comment by 95-martin-orion** [comment]
Reviewers can focus on `json_serialization[_test].py` - the remaining files just bring the rest of Cirq in line with this change.

**Comment by mpharrigan** [comment]
wow you went for it

File: cirq-core/cirq/protocols/json_serialization.py [hunk 1]
```diff
@@ -249,7 +269,24 @@ def dataclass_json_dict(obj: Any, namespace: str = None) -> Dict[str, Any]:
...
+    base_dict = obj._json_dict_()
+    if 'cirq_type' in base_dict:
+        # TODO: upgrade to ValueError in v0.15
+        warnings.warn(
+            f"Found 'cirq_type': '{base_dict['cirq_type']}' in _json_dict_. "
+            f"Custom values of this field are not permitted, and will produce "
+            "an error starting in Cirq v0.15.",
+            DeprecationWarning,
+        )
+    return {'cirq_type': json_cirq_type(type(obj)), **base_dict}
```
Comment (mpharrigan) [inline cirq-core/cirq/protocols/json_serialization.py:289]: would this work with `json_cirq_type(obj)`? or does it need a type? pros and cons of handling this here vs in `json_cirq_type`?

**Review by mpharrigan — APPROVED** [review]
Looks very comprehensive

**Comment by mpharrigan** [comment]
Right now, the magic `_from_json_dict_` always gets passed the `cirq_type` string. I think most implementations ignore it. Do you see this part of the json stack being modified in the future?

File: cirq-core/cirq/protocols/json_serialization.py (hunk 1 above)
Comment (95-martin-orion) [reply cirq-core/cirq/protocols/json_serialization.py:289]: I'm pretty sure I could make `json_cirq_type` accept both objects and types - we'd just need to differentiate between the two inside the method. However, the only case I can think of where a user needs to call `json_cirq_type` directly is for type-serialization, in which case they ought to be passing in types anyways.

**Comment by 95-martin-orion** [comment]
> Right now, the magic `_from_json_dict_` always gets passed the `cirq_type` string. I think most implementations ignore it. Do you see this part of the json stack being modified in the future?

AFAIK `cirq_type` is only used for selecting the type to deserialize into, especially now that `json_cirq_type` can be used to discover the `cirq_type` of a class. That said, I don't see any strong incentive to remove/change it.

**Comment by mpharrigan** [comment]
Yeah, I don't think we should change it. I think the motivation was for backwards compatibility support: if you merge what used to be two different classes into one type where the distinction is made via a member variable or something, you could have two "cirq_types" pointing to the same class and it could deal with it.

**Pull request merged**